src/inmanta/deploy/scheduler.py

```diff
@@ -848,46 +1025,54 @@ async def get_resource_intent(self, resource: ResourceIdStr) -> Optional[Resourc
         async with self._scheduler_lock:
             # fetch resource details under lock
             resource_details = self._get_resource_intent(resource)
             if resource_details is None or self._state.resource_state[resource].blocked is BlockedStatus.YES:
+                # We are trying to deploy a stale resource.
                 return None
             dependencies = await self._get_last_non_deploying_state_for_dependencies(resource=resource)
             self._deploying_latest.add(resource)
-            return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=dependencies)
+            resource_intent = ResourceIntent(
+                model_version=self._state.version, details=resource_details, dependencies=dependencies
+            )
+            # Update the state in the database.
+            await self.send_in_progress(action_id, Id.parse_id(ResourceVersionIdStr(f"{resource},v={self._state.version}")))
+            return resource_intent
 
-    async def deploy_done(
-        self,
-        resource: ResourceIdStr,
-        *,
-        attribute_hash: str,
-        resource_state: const.HandlerResourceState,
+    async def deploy_done(self, attribute_hash: str, result: DeployResult) -> None:
+        deployment_result: DeploymentResult = DeploymentResult.from_handler_resource_state(result.resource_state)
+        try:
+            await self.update_scheduler_state_for_finished_deploy(attribute_hash, result, deployment_result)
+        except StaleResource:
+            # The resource is no longer managed, no need to update the database state.
+            pass
+        else:
+            # Write deployment result to the database.
+            await self.send_deploy_done(attribute_hash, result, deployment_result)
+
+    async def update_scheduler_state_for_finished_deploy(
+        self, attribute_hash: str, result: DeployResult, deployment_result: DeploymentResult
     ) -> None:
-        # Translate deploy result status to the new deployment result state
-        deployment_result: DeploymentResult
-        match resource_state:
-            case const.HandlerResourceState.deployed:
-                deployment_result = DeploymentResult.DEPLOYED
-            case const.HandlerResourceState.skipped | const.HandlerResourceState.skipped_for_dependency:
-                deployment_result = DeploymentResult.SKIPPED
-            case _:
-                deployment_result = DeploymentResult.FAILED
-
-        status = (
-            ComplianceStatus.COMPLIANT if deployment_result is DeploymentResult.DEPLOYED else ComplianceStatus.NON_COMPLIANT
-        )
+        """
+        Update the state of the scheduler based on the DeploymentResult of the given resource.
+
+        :raise StaleResource: This update is about a resource that is no longer managed by the server.
+        """
+        resource_id: ResourceIdStr = result.resource_id
```

Comment:
Nice, I hadn't considered that this contains the resource id.

Reply:
Tiny preference to keep using `resource` for the name though. It's what the rest of this module does.